src/preact/context.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createContext, createElement} from 'preact';
+import {useContext} from 'preact/hooks';
+
+/**
+ * @typedef {{
+ *   Provider: *,
+ *   Consumer: *,
+ * }}
+ */
+let ContextDef;
+
+/** @type {ContextDef} */
+let context;
+
+/**
+ * The external context given to React components to control whether they can
+ * render/play/etc.
+ *
+ * - renderable: whether this vDOM area is renderable. Analogous to
+ *   `display-locking` CSS.
+ * - playable: whether the playback is allowed in this vDOM area. If playback
+ *   is not allow, the component must immediately stop the playback.
+ *
+ * @return {!ContextDef}
+ */
+export function getAmpContext() {
+  return (
+    context ||
+    (context = createContext({
```

Comment:
Would it make sense for us to do an exception for side-effects in this case? I'm thinking this class is part of the AMP's React base that could be exportable as it's own library. So the closer we could get to the canonical state for React the better for these modules.

Reply:
For use internally, this will be no different. If we export it externally, we could do a code mod? But either way, they'll be different context instances in different modules, and changing it doesn't help.

Reply:
Yeah. So I'm wondering why not just allow side-effects in this case to make this more-or-less a classical Preact/React-style declaration?